clustermesh-apiserver/coredns-mcsapi-auto-configure/cmd.go

```diff
@@ -0,0 +1,163 @@
+		Run: func(cmd *cobra.Command, args []string) {
+			// slogloggercheck: it has been initialized in the PreRun function.
+			if err := h.Run(logging.DefaultSlogLogger); err != nil {
+				// slogloggercheck: log fatal errors using the default logger before it's initialized.
+				logging.Fatal(logging.DefaultSlogLogger, err.Error())
+			}
+		},
+		PreRun: func(cmd *cobra.Command, args []string) {
+			// Overwrite the metrics namespace with the one specific for KVStoreMesh
+			option.Config.SetupLogging(h.Viper(), "coredns-mcsapi-auto-configure")
+
+			// slogloggercheck: the logger has been initialized in the SetupLogging call above
+			log := logging.DefaultSlogLogger.With(logfields.LogSubsys, "coredns-mcsapi-auto-configure")
+
+			option.Config.Populate(log, h.Viper())
+			option.LogRegisteredSlogOptions(h.Viper(), log)
+			log.Info("Cilium MCS-API CoreDNS auto configuration", logfields.Version, version.Version)
+		},
+	}
+
+	h.RegisterFlags(rootCmd.Flags())
+	rootCmd.AddCommand(h.Command())
+
+	return rootCmd
+}
+
+func configureCoreDNS(jg job.Group, logger *slog.Logger, shutdowner hive.Shutdowner, client client.Clientset, config coreDNSConfig) {
+	jg.Add(job.OneShot("clustermesh-nodemanager-notifier", func(ctx context.Context, _ cell.Health) (err error) {
+		logger.Info("Starting MCS-API CoreDNS auto configuration")
+		defer func() {
+			shutdowner.Shutdown(hive.ShutdownWithError(err))
+		}()
+
+		var coreDNSConfigmap *corev1.ConfigMap
+		coreDNSConfigmap, err = client.CoreV1().ConfigMaps(config.CoreDNSNamespace).Get(ctx, config.CoreDNSConfigMapName, metav1.GetOptions{})
+		if err != nil {
+			logger.Error("Failed to get CoreDNS ConfigMap", logfields.Error, err)
+			return
+		}
+
+		var corefile string
+		var found bool
+		if corefile, found = coreDNSConfigmap.Data["Corefile"]; !found {
+			err = fmt.Errorf("Corefile not found in ConfigMap %s/%s\n", config.CoreDNSNamespace, config.CoreDNSConfigMapName)
+			logger.Error(err.Error())
+			return
+		}
+
+		if corefile, err = updateCorefile(logger, config.CoreDNSClusterDomain, config.CoreDNSClustersetDomain, corefile); err != nil {
+			return
+		}
+		if corefile == "" {
+			return
+		}
+		coreDNSConfigmap.Data["Corefile"] = corefile
+
+		if _, err = client.CoreV1().ConfigMaps(config.CoreDNSNamespace).Update(ctx, coreDNSConfigmap, metav1.UpdateOptions{}); err != nil {
+			logger.Error("Failed to update CoreDNS ConfigMap", logfields.Error, err)
+			return
+		}
+
+		logger.Info("CoreDNS ConfigMap was updated successfully")
+
+		err = restartCoreDNS(ctx, logger, client, config)
+		if err != nil {
+			logger.Info("CoreDNS is rolling out with the new configuration")
+		}
+		return
+	}))
+}
+
+func updateCorefile(logger *slog.Logger, clusterDomain, clustersetDomain string, corefile string) (string, error) {
+	if strings.Contains(corefile, clustersetDomain) || strings.Contains(corefile, "multicluster") {
+		logger.Warn("CoreDNS might already have MCS-API configuration, skipping configuration")
```

Comment:
I wonder if `level=warn` is the right choice here. What happens it the user does a Cilium upgrade via `helm upgrade` - will the job be re-triggered and it it always emit a warning? In that case, I think it might be better to leave this at `level=info`